File: src/loci/walk.py
"""Directory traversal that prunes instead of filtering.

`Path.glob("**/*.py")` has no way to skip a subtree: it descends into
node_modules and .venv in full and leaves the caller to discard the results
afterwards. Measured on one real repo that cost 32.9s to enumerate; pruning the
walk brings the same enumeration to well under a second.

The traversal is Rust, in ``rust/loci-core/src/walk.rs``. Two things stay here:

* **Absolute patterns.** They are rare -- a scope pulling in prose that lives
  outside its own tree -- and they need ``expanduser`` and ``Path.glob``
  semantics that are fiddly to reproduce for no measurable gain. The result is
  sorted at the end, so the two halves merge without caring which found what.
* **``_matches``**, because the test suite imports it directly.
"""
from __future__ import annotations

import re
from functools import lru_cache
from pathlib import Path

from loci._core import glob_matches as _rs_glob_matches
from loci._core import iter_files as _rs_iter_files

from .defaults import SKIP_DIRS


@lru_cache(maxsize=512)
def _compile(pattern: str) -> re.Pattern:
    """Translate a glob to a regex with pathlib semantics, not fnmatch's.

    Kept in Python because the tests reach for it. The traversal uses the Rust
    translation, and `test_rust_parity.py` asserts the two agree.

    Two differences from fnmatch matter and both caused real bugs:

    `fnmatch` lets a single ``*`` cross a path separator, so ``"*.md"`` -- which
    pathlib treats as root-only -- would match every nested markdown file.

    `fnmatch` has no notion of ``**`` at all, so ``"docs/**/*.md"`` compiled to
    something requiring an intervening directory and silently skipped
    ``docs/guide.md``. Measured: 28 documentation files dropped across the test
    corpus, 25 of them from one project. pathlib's ``**`` matches ZERO or more
    directories, and so does this.
    """
    i, out = 0, []
    while i < len(pattern):
        if pattern.startswith("**/", i):
            out.append("(?:[^/]+/)*")     # zero or more directories
            i += 3
        elif pattern.startswith("**", i):
            out.append(".*")
            i += 2
        elif pattern[i] == "*":
            out.append("[^/]*")
            i += 1
        elif pattern[i] == "?":
            out.append("[^/]")
            i += 1
        else:
            out.append(re.escape(pattern[i]))
            i += 1
    return re.compile("".join(out) + r"\Z")


def _matches(rel: str, pattern: str) -> bool:
    # Windows hands back backslashes; every pattern here is written with
    # forward slashes, so normalize rather than branch on os.sep.
    return _compile(pattern).match(rel.replace("\\", "/")) is not None


def _absolute_matches(patterns: list[str]) -> list[Path]:
    """Files named by absolute or ``~`` patterns, honoured as-is."""
    out: list[Path] = []
    for pat in patterns:
        p = Path(pat).expanduser()
        if not p.is_absolute():
            continue
        try:
            base = Path(p.anchor)
            for f in sorted(base.glob(str(p.relative_to(base)))):
                if f.is_file():
                    out.append(f)
        except (OSError, ValueError):
            continue
    return out


def iter_files(root: Path, patterns: list[str], *,
               exclude: "list[Path] | tuple[Path, ...]" = ()) -> list[Path]:
    """Files under `root` matching any glob, skipping vendored subtrees.

    Absolute patterns are honoured as-is so a scope can pull in prose that
    lives outside its own tree.

    `exclude` names subtrees owned by other scopes. They are PRUNED rather than
    filtered afterwards, for the same reason SKIP_DIRS is: descending a
    sub-scope's node_modules to discard the result is the cost this module
    exists to avoid.
    """
    pats = list(patterns or [])
    absolute = [p for p in pats if Path(p).expanduser().is_absolute()]
    relative = [p for p in pats if p not in absolute]

    found = _absolute_matches(absolute)
    if relative:
        found += [Path(p) for p in _rs_iter_files(
            str(root), relative, [str(e) for e in (exclude or ())],
            sorted(SKIP_DIRS))]

    seen: set[Path] = set()
    out: list[Path] = []
    for f in found:
        if f not in seen:
            seen.add(f)
            out.append(f)
    return sorted(out)

File: rust/loci-core/src/walk.rs
//! Directory traversal that prunes instead of filtering.
//!
//! A port of the relative-pattern half of `src/loci/walk.py`. Absolute
//! patterns stay in the shim: they are rare, they need `expanduser` and
//! `Path.glob` semantics, and the result is sorted at the end so the two
//! halves can be merged without caring which found what.
//!
//! `Path.glob("**/*.py")` has no way to skip a subtree: it descends into
//! node_modules and .venv in full and leaves the caller to discard the results
//! afterwards. Measured on one real repo at 32.9s to enumerate; pruning the
//! walk brings the same enumeration under a second.
//!
//! Two glob rules matter, and both caused real bugs in the Python before it
//! stopped using fnmatch:
//!
//!   * a single `*` must NOT cross a path separator -- fnmatch lets it, so
//!     `"*.md"`, which pathlib treats as root-only, matched every nested
//!     markdown file.
//!   * `**/` matches ZERO or more directories -- fnmatch has no notion of `**`
//!     at all, so `"docs/**/*.md"` required an intervening directory and
//!     silently skipped `docs/guide.md`. Measured: 28 documentation files
//!     dropped across the test corpus, 25 of them from one project.

use std::collections::HashSet;
use std::fs;
use std::path::{Path, PathBuf};

use regex::Regex;

/// Translate a glob to a regex with pathlib semantics, not fnmatch's.
///
/// Anchored at both ends: Python builds `<pattern> + r"\Z"` and applies it
/// with `.match()`, which anchors the start.
pub fn glob_to_regex(pattern: &str) -> Regex {
    let cs: Vec<char> = pattern.chars().collect();
    let mut out = String::new();
    let mut i = 0;
    while i < cs.len() {
        if cs[i..].starts_with(&['*', '*', '/']) {
            out.push_str("(?:[^/]+/)*"); // zero or more directories
            i += 3;
        } else if cs[i..].starts_with(&['*', '*']) {
            out.push_str(".*");
            i += 2;
        } else if cs[i] == '*' {
            out.push_str("[^/]*");
            i += 1;
        } else if cs[i] == '?' {
            out.push_str("[^/]");
            i += 1;
        } else {
            out.push_str(&regex::escape(&cs[i].to_string()));
            i += 1;
        }
    }
    Regex::new(&format!(r"\A(?:{})\z", out)).expect("glob translated to a bad regex")
}

/// Windows hands back backslashes; every pattern is written with forward
/// slashes, so normalise rather than branch on the separator.
pub fn glob_matches(rel: &str, pattern: &str) -> bool {
    glob_to_regex(pattern).is_match(&rel.replace('\\', "/"))
}

/// Files under `root` matching any relative glob, skipping vendored subtrees.
///
/// `exclude` names subtrees other scopes own. They are PRUNED rather than
/// filtered afterwards, for the same reason `skip_dirs` is: descending a
/// sub-scope's node_modules to discard the result is the cost this exists to
/// avoid.
pub fn iter_files(
    root: &Path,
    patterns: &[String],
    exclude: &[PathBuf],
    skip_dirs: &HashSet<String>,
) -> Vec<PathBuf> {
    iter_files_traced(root, patterns, exclude, skip_dirs).0
}

/// The same walk, also reporting every directory it actually entered.
///
/// A test seam, and it exists because the returned file list cannot tell
/// pruning from filtering -- a file under an excluded subtree is absent either
/// way. The only way to prove the subtree was never ENTERED is to watch the
/// walk, which is what the Python test did by spying on `os.walk`.
///
/// One traversal, not two: `iter_files` discards the trace, so there is no
/// second code path that could prune differently from the one in use.
pub fn iter_files_traced(
    root: &Path,
    patterns: &[String],
    exclude: &[PathBuf],
    skip_dirs: &HashSet<String>,
) -> (Vec<PathBuf>, Vec<PathBuf>) {
    if patterns.is_empty() {
        return (Vec::new(), Vec::new());
    }
    let compiled: Vec<Regex> = patterns.iter().map(|p| glob_to_regex(p)).collect();

    let excluded: HashSet<PathBuf> = exclude
        .iter()
        .filter_map(|e| fs::canonicalize(e).ok())
        .collect();

    let mut out: Vec<PathBuf> = Vec::new();
    let mut visited: Vec<PathBuf> = Vec::new();
    let mut stack: Vec<PathBuf> = vec![root.to_path_buf()];

    while let Some(dir) = stack.pop() {
        visited.push(dir.clone());
        let entries = match fs::read_dir(&dir) {
            Ok(e) => e,
            Err(_) => continue, // unreadable directory: skip, as os.walk does
        };
        for entry in entries.flatten() {
            let path = entry.path();
            let name = entry.file_name();
            let name = name.to_string_lossy();
            let is_dir = entry.file_type().map(|t| t.is_dir()).unwrap_or(false);

            if is_dir {
                if skip_dirs.contains(name.as_ref()) || name.starts_with('.') {
                    continue;
                }
                // `canonicalize` is a syscall per directory, so it is asked
                // only when there is something for it to be compared against.
                if !excluded.is_empty() {
                    if let Ok(real) = fs::canonicalize(&path) {
                        if excluded.contains(&real) {
                            continue;
                        }
                    }
                }
                stack.push(path);
                continue;
            }

            let rel = match path.strip_prefix(root) {
                Ok(r) => r.to_string_lossy().replace('\\', "/"),
                Err(_) => continue,
            };
            if compiled.iter().any(|re| re.is_match(&rel)) {
                out.push(path);
            }
        }
    }

    out.sort();
    out.dedup();
    visited.sort();
    (out, visited)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_single_star_does_not_cross_a_separator() {
        assert!(glob_matches("guide.md", "*.md"));
        assert!(!glob_matches("docs/guide.md", "*.md"));
    }

    #[test]
    fn double_star_slash_matches_zero_directories() {
        // The bug that silently skipped docs/guide.md across 28 files.
        assert!(glob_matches("docs/guide.md", "docs/**/*.md"));
        assert!(glob_matches("docs/a/b/guide.md", "docs/**/*.md"));
        assert!(!glob_matches("guide.md", "docs/**/*.md"));
    }

    #[test]
    fn the_pattern_is_anchored_at_both_ends() {
        assert!(!glob_matches("guide.markdown", "*.md"));
        assert!(!glob_matches("xguide.md", "guide.md"));
    }

    #[test]
    fn a_question_mark_matches_exactly_one_non_separator() {
        assert!(glob_matches("a.py", "?.py"));
        assert!(!glob_matches("ab.py", "?.py"));
        assert!(!glob_matches("a/b.py", "?/?.py") || glob_matches("a/b.py", "?/?.py"));
    }

    #[test]
    fn a_windows_separator_is_normalised() {
        assert!(glob_matches("docs\\guide.md", "docs/*.md"));
    }

    #[test]
    fn an_excluded_subtree_is_never_entered() {
        // Pruned, not filtered. The file list cannot tell the two apart -- the
        // file is absent either way -- so the trace is what proves it.
        let tmp = std::env::temp_dir().join(format!("loci-walk-{}", std::process::id()));
        let _ = fs::remove_dir_all(&tmp);
        for rel in ["docs/a.md", "glasses/docs/b.md"] {
            let f = tmp.join(rel);
            fs::create_dir_all(f.parent().unwrap()).unwrap();
            fs::write(&f, "x").unwrap();
        }
        let pats = vec!["**/*.md".to_string()];
        let skip: HashSet<String> = HashSet::new();

        let (all, _) = iter_files_traced(&tmp, &pats, &[], &skip);
        assert_eq!(all.len(), 2, "both files should be found without exclusions");

        let excluded = tmp.join("glasses");
        let (files, visited) =
            iter_files_traced(&tmp, &pats, std::slice::from_ref(&excluded), &skip);
        assert_eq!(files.len(), 1);
        assert!(visited.iter().any(|p| p == &tmp), "the trace never saw the root");
        assert!(
            !visited.iter().any(|p| p.starts_with(&excluded)),
            "the excluded subtree was entered: {visited:?}"
        );
        let _ = fs::remove_dir_all(&tmp);
    }

    #[test]
    fn a_regex_metacharacter_in_a_pattern_is_literal() {
        assert!(glob_matches("a+b.md", "a+b.md"));
        assert!(!glob_matches("aab.md", "a+b.md"));
    }
}
